python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,529 @@
+
+        if isinstance(key, str):
+            if key == "input":
+                return self.input_bindings
+            if key == "output":
+                return self.output_bindings
+
+        raise RuntimeError(f"{key} not found.")
+
+    def get_config(self):
+        """ Get configuration information in dictionary form."""
+
+        # Use topological sort to get correct order of modules.
+        self.dag_topology_sort()
+        mconfig = {}
+        for mod in self.mod_wrapper:
+            # Generate pipeline configure.
+            mconf = {}
+            output_conf = []
+            module = self.mod_wrapper[mod]
+            for _, binding in module.output_bindings.bindings.items():
+                dep_conf = []
+                output = {}
+                if binding.bindings:
+                    for dep in binding.bindings:
+                        dep_item = {}
+                        _, dname = dep.get_name()
+                        dep_item["mod_idx"] = dep.get_owner_idx()
+                        dep_item["input_name"] = dname
+                        dep_conf.append(dep_item)
+
+                # Ouput_idx start from 0.
+                output["output_idx"] = int(binding.name)
+                output["dependent"] = dep_conf
+                output_conf.append(output)
+
+            mconf["mod_idx"] = module.idx
+            mconf["output"] = output_conf
+
+            # Build configuration with parameters.
+            mconfig[mod] = {
+                "pipeline": mconf,
+                "target_host": module.target_host,
+                "mod_name": "default",
+                "build": module.build_func,
+                "params": module.params,
+                "target": module.target,
+                "dev": module.dev,
+            }
+
+        return mconfig
+
+    def dag_topology_sort(self):
+        """ Use topological sort to get order of pipeline modules."""
+        mlist = []
+        mod_wrapper = self.mod_wrapper.copy()
+        while mod_wrapper:
+            temp_list = []
+            for mod, wrapper in mod_wrapper.items():
+                if wrapper.is_root_mod():
+                    temp_list.append(mod)
+                    wrapper.remove_self_from_bindings()
+
+            for mod in temp_list:
+                mod_wrapper.pop(mod, None)
+
+            mlist += temp_list
+
+        for mod, i in zip(mlist, range(len(mlist))):
+            self.mod_wrapper[mod].set_idx_name(i + 1)
+
+    def get_mod_idx(self, mod):
+        """Return module index."""
+        idx = self.mod_wrapper[mod].idx
+        return idx
+
+    def pipe_input(self, name):
+        """Return the corresponding input binding interface accordding the name."""
+        return self.input_bindings[name]
+
+    def pipe_output(self, idx):
+        """Return the corresponding output binding interface accordding the name."""
+        return self.output_bindings[idx]
+
+
+class PipelineExecutorFactoryModule(object):
```

Comment:
I commented this before, this is not a factory class.

Reply:
moved the pipeline executor create logic into PipelineExecutorFactoryModule, now PipelineExecutorFactoryModule response for create and PipelineModule response to expose functions like setinput etc to caller.

Reply:
See https://github.com/apache/tvm/pull/8702/files#r707216451

There is no reason this needs to be a class. Replace it with a function.

Reply:
The benefit to make it to be a class is that we can decouple  executor maintain logic with executor function export logic, for example after this initial version, we can add a export function to export library and configuration, into PipelineExecutorFactoryModule, the class structure can make such logic be more clear, 

at same time PipelineExecutorFactoryModule follow same logic like GraphExecutorFactoryModule that used by GraphExecutor, doing so also can make code logic more consistent. please let me know how you think.

Reply:
ok makes sense